addons/product/models/product.py

```diff
@@ -422,6 +422,40 @@ def unlink(self):
         self.clear_caches()
         return res
 
+    @api.multi
+    def _unlink_or_archive(self):
+        """Unlink or archive products.
+        Try in batch as much as possible because it is much faster.
+        Use dichotomy when an exception occurs.
+        """
+
+        # Avoid access errors in case the products is shared amongst companies
+        # but the underlying objects are not. If unlink fails because of an
+        # AccessError (e.g. while recomputing fields), the 'write' call will
+        # fail as well for the same reason since the field has been set to
+        # recompute.
+        self.check_access_rights('unlink')
+        self.check_access_rule('unlink')
+        self.check_access_rights('write')
+        self.check_access_rule('write')
+        self = self.sudo()
+
+        try:
+            with self.env.cr.savepoint(), tools.mute_logger('odoo.sql_db'):
+                self.unlink()
+        except Exception:
+            # We catch all kind of exceptions to be sure that the operation
+            # doesn't fail.
+            if len(self) > 1:
+                self[:len(self) // 2]._unlink_or_archive()
+                self[len(self) // 2:]._unlink_or_archive()
```

Comment:
@antonylesuisse 

your pov ?

Reply:
For the following test case, realistic enough:

- create a template
- give it 3 attributes, 10 values each
- -> that generates 1000 variants
- add the first variant to an order to test it
- not happy with the config, remove one of the attributes from the template
- -> that removes the 1000 variants

This is slow because of the `savepoint` used to archive if `unlink` fails.

With a loop (one by one) it takes 5min.
With dichotomy it takes 14s.

I added this result on the commit message.